File: models/gene.py
import random
from config import ALL_SUBSTANCE_NAMES
from models.action import Action
from models.trigger import Trigger


class Gene:
    """
    Базовый ген — связывает триггер и действие.
    Если условие триггера выполняется, то активируется действие.
    """

    def __init__(
        self,
        receptor: str,
        trigger: Trigger,
        action: Action,
        active: bool = True,
        mutation_rate: float = 0.07
    ):
        """
        :param receptor: имя вещества или параметра, на который реагирует ген
        :param trigger: условие активации
        :param action: действие при активации
        :param active: активен ли ген (может быть выключен)
        """
        self.receptor = receptor
        self.trigger = trigger
        self.action = action
        self.active = active
        self.mutation_rate = mutation_rate


    def try_activate(self, cell: "Cell", environment: "Environment"):
        """Проверяет условие и выполняет действие, если сработал триггер."""
        if not self.active:
            return

        value = None

        # --- 1. Попробуем получить значение вещества из среды ---
        if self.receptor not in ("energy", "health"):
            x, y = cell.get_int_position()
            substance = environment.grid.get_substance(x, y, self.receptor)
            if substance:
                value = substance.concentration

        # --- 2. Если вещество не найдено, пробуем взять из клетки ---
        if value is None:
            value = getattr(cell, self.receptor, None)

        if value is None:
            return

        if self.trigger.check(value):
            self.action.execute(cell, environment)


    def mutate(self):
        """Простая мутация параметров гена."""
        if not self.is_triggered_mutation():
            return None

        if self.is_triggered_mutation():
            self.active = not self.active

        if self.is_triggered_mutation():
            if self.receptor in ("energy", "health"):
                self.trigger.threshold = random.uniform(1, 100.0)
            else:
                self.trigger.threshold = random.uniform(0.1, 10.0)

        if self.is_triggered_mutation():
            self.action.power = random.uniform(0.1, 10.0)

        if self.is_triggered_mutation():
            return Gene.create_random_gene()

        if self.is_triggered_mutation():
            self.action.move_mode = random.choice([
                Action.MOVE_RANDOM,
                Action.MOVE_TOWARD,
                Action.MOVE_AWAY,
                Action.MOVE_AROUND,
                None,
            ])

        if self.is_triggered_mutation():
            self.mutation_rate *= random.choice((1.15, 0.85))
            self.mutation_rate = min(self.mutation_rate, 1.0)

        return None

    def is_triggered_mutation(self):
        return random.random() < self.mutation_rate

    @classmethod
    def create_random_gene(cls) -> 'Gene':
        """Создаёт случайный ген."""
        # 85% генов реагируют на вещества, 15% — на внутренние параметры клетки
        if random.random() < 0.85:
            receptor = random.choice(ALL_SUBSTANCE_NAMES)
        else:
            receptor = random.choice(["energy", "health"])

        if receptor in ("energy", "health"):
            threshold = random.uniform(1, 100.0)
        else:
            threshold = random.uniform(0.1, 10.0)
        mode = random.choice((Trigger.LESS, Trigger.GREATER))
        trigger = Trigger(threshold, mode)

        action_type = random.choice((
            Action.DIVIDE, Action.EMIT, Action.ABSORB,
            Action.MOVE, Action.HEALS
        ))

        if action_type == Action.MOVE:
            move_mode = random.choice([
                Action.MOVE_RANDOM,
                Action.MOVE_TOWARD,
                Action.MOVE_AWAY,
                Action.MOVE_AROUND,
            ])
            substance_name = random.choice(ALL_SUBSTANCE_NAMES)
        else:
            move_mode = None
            substance_name = random.choice(ALL_SUBSTANCE_NAMES)

        action = Action(
            type_=action_type,
            power=random.uniform(0.1, 10.0),
            substance_name=substance_name,
            move_mode=move_mode,
        )

        return Gene(
            receptor=receptor,
            trigger=trigger,
            action=action
        )

    def to_tuple(self) -> tuple:
        """Простой стабильный ключ гена."""
        return (
            self.receptor,
            getattr(self.trigger, "mode", ""),
            round(getattr(self.trigger, "threshold", 0.0), 3),
            getattr(self.action, "type", ""),
            round(getattr(self.action, "power", 0.0), 3),
            getattr(self.action, "substance_name", "") or "",
            getattr(self.action, "move_mode", "") or "",
            self.active,
        )


    def clone(self) -> 'Gene':
        """Создаёт копию без мутации."""
        return Gene.from_dict(self.to_dict())

    def to_dict(self):
        return {
            "receptor": self.receptor,
            "trigger": self.trigger.to_dict(),
            "action": self.action.to_dict(),
            "active": self.active,
            "mutation_rate": self.mutation_rate,
        }

    @classmethod
    def from_dict(cls, data):
        return cls(
            receptor=data["receptor"],
            trigger=Trigger.from_dict(data["trigger"]),
            action=Action.from_dict(data["action"]),
            active=data.get("active", True),
            mutation_rate=data.get("mutation_rate", 0.07)
        )

    def __repr__(self):
        return (f"Gene(receptor={self.receptor}, {self.trigger}, "
                f"{self.action}, active={self.active})")

File: java/evolution/src/main/java/com/life/evolution/model/Gene.java
package com.life.evolution.model;

import com.life.evolution.config.Config;

import java.util.*;

public class Gene {

    private String receptor;           // substance or parameter name
    private Trigger trigger;           // activation condition
    private Action action;             // action to execute
    private boolean active;            // gene enabled/disabled
    private double mutationRate;       // probability of mutation

    public Gene(String receptor, Trigger trigger, Action action,
                boolean active, double mutationRate) {

        this.receptor = receptor;
        this.trigger = trigger;
        this.action = action;
        this.active = active;
        this.mutationRate = mutationRate;
    }

    public Gene(String receptor, Trigger trigger, Action action) {
        this(receptor, trigger, action, true, 0.07);
    }

    // --------------------------------------------------------
    // EXECUTION
    // --------------------------------------------------------
    public void tryActivate(Cell cell, Environment env) {
        if (!active) return;

        Double value = null;

        // Try reading substance from environment grid
        if (!Objects.equals(receptor, "energy") &&
                !Objects.equals(receptor, "health")) {

            int[] pos = cell.getIntPosition();
            Substance s = env.getGrid().getSubstance(pos[0], pos[1], receptor);

            if (s != null) value = s.getConcentration();
        }

        // fallback to cell field
        if (value == null) {
            value = switch (receptor) {
                case "energy" -> cell.getEnergy();
                case "health" -> cell.getHealth();
                default -> null;
            };
        }

        if (value == null) return;

        if (trigger.check(value)) {
            action.execute(cell, env);
        }
    }

    // --------------------------------------------------------
    // MUTATION
    // --------------------------------------------------------
    public Gene mutate() {

        if (isTriggeredMutation()) {
            active = !active;
        }

        if (isTriggeredMutation()) {
            receptor = Config.randomSubstanceName();
        }

        if (isTriggeredMutation()) {
            if (receptor.equals("energy") || receptor.equals("health")) {
                trigger.setThreshold(randomDouble(1, 100));
            } else {
                trigger.setThreshold(randomDouble(0.1, 10));
            }
        }

        if (isTriggeredMutation()) {
            action.setPower(randomDouble(0.1, 10));
        }

        if (isTriggeredMutation()) {
            return Gene.createRandomGene();
        }

        if (isTriggeredMutation()) {
            action.setMoveMode(randomChoice(List.of(
                    Action.MOVE_RANDOM,
                    Action.MOVE_TOWARD,
                    Action.MOVE_AWAY,
                    Action.MOVE_AROUND,
                    null
            )));
        }

        if (isTriggeredMutation()) {
            mutationRate *= randomChoice(List.of(1.15, 0.85));
            mutationRate = Math.min(mutationRate, 1.0);
        }

        return null;
    }

    private boolean isTriggeredMutation() {
        return Math.random() < mutationRate;
    }

    // --------------------------------------------------------
    // RANDOM GENE FACTORY
    // --------------------------------------------------------
    public static Gene createRandomGene() {

        String receptor;

        if (Math.random() < 0.85) {
            receptor = Config.randomSubstanceName();
        } else {
            receptor = randomChoice(List.of("energy", "health"));
        }

        double threshold;
        if (receptor.equals("energy") || receptor.equals("health")) {
            threshold = randomDouble(1, 100);
        } else {
            threshold = randomDouble(0.1, 10);
        }

        String mode = randomChoice(List.of(Trigger.LESS, Trigger.GREATER));
        Trigger trig = new Trigger(threshold, mode);

        String actionType = randomChoice(List.of(
                Action.DIVIDE,
                Action.EMIT,
                Action.ABSORB,
                Action.MOVE,
                Action.HEALS
        ));

        String moveMode = null;
        String substanceName;

        if (actionType.equals(Action.MOVE)) {
            moveMode = randomChoice(List.of(
                    Action.MOVE_RANDOM,
                    Action.MOVE_TOWARD,
                    Action.MOVE_AWAY,
                    Action.MOVE_AROUND
            ));
            substanceName = Config.randomSubstanceName();
        } else {
            substanceName = Config.randomSubstanceName();
        }

        Action action = new Action(
                actionType,
                randomDouble(0.1, 10),
                substanceName,
                moveMode
        );

        return new Gene(receptor, trig, action);
    }

    // --------------------------------------------------------
    // JSON SERIALIZATION
    // --------------------------------------------------------
    public Map<String, Object> toMap() {
        Map<String, Object> m = new HashMap<>();
        m.put("receptor", receptor);
        m.put("trigger", trigger.toMap());
        m.put("action", action.toMap());
        m.put("active", active);
        m.put("mutation_rate", mutationRate);
        return m;
    }

    @SuppressWarnings("unchecked")
    public static Gene fromMap(Map<String, Object> m) {
        return new Gene(
                (String) m.get("receptor"),
                Trigger.fromMap((Map<String, Object>) m.get("trigger")),
                Action.fromMap((Map<String, Object>) m.get("action")),
                (boolean) m.getOrDefault("active", true),
                ((Number) m.getOrDefault("mutation_rate", 0.07)).doubleValue()
        );
    }

    // --------------------------------------------------------
    // SIGNATURE (used for color)
    // --------------------------------------------------------
    public String toTuple() {
        return String.join("|",
                receptor,
                trigger.getMode(),
                String.format("%.3f", trigger.getThreshold()),
                action.getType(),
                String.format("%.3f", action.getPower()),
                action.getSubstanceName() != null ? action.getSubstanceName() : "",
                action.getMoveMode() != null ? action.getMoveMode() : "",
                Boolean.toString(active)
        );
    }

    // --------------------------------------------------------
    // UTIL
    // --------------------------------------------------------
    private static double randomDouble(double a, double b) {
        return Math.random() * (b - a) + a;
    }

    private static <T> T randomChoice(List<T> items) {
        return items.get((int) (Math.random() * items.size()));
    }

    // --------------------------------------------------------
    // GETTERS
    // --------------------------------------------------------
    public String getReceptor() { return receptor; }
    public Trigger getTrigger() { return trigger; }
    public Action getAction() { return action; }
    public boolean isActive() { return active; }
    public double getMutationRate() { return mutationRate; }
}
